config/validate_test.go

```diff
@@ -0,0 +1,212 @@
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		err := os.Unsetenv(ConfigEnvironmentVariable)
+		require.NoError(t, err)
+	})
+
+	parsedConfig, err := FromEnv()
+	require.NoError(t, err)
+	require.Equal(t, expectedConfig, parsedConfig)
+}
+
+func OpenFile(t *testing.T, path string) string {
+	byteFile, err := ioutil.ReadFile(path)
+	require.NoError(t, err)
+	return string(byteFile)
+}
+
+var (
+	expectedConfig = &Config{
+		Secret: AclTokenSecret{
+			Provider: "secrets-manager",
+			Configuration: SecretConfiguration{
+				Prefix:                     "asdf",
+				ConsulClientTokenSecretARN: "ARN",
+			},
+		},
+		Mesh: Mesh{
+			Service: ServiceRegistration{
+				Name: "blah",
+				Port: 1234,
+				Tags: []string{"tag1"},
+				Meta: map[string]string{"a": "1"},
+			},
+			Proxy: &AgentServiceConnectProxyConfig{
+				Upstreams: []Upstream{
+					{
+						DestinationName: "asdf",
+						LocalBindPort:   543,
+					},
+				},
+			},
+			HealthSyncContainers: []string{"container1"},
+			BootstrapDir:         "/consul/",
+		},
+	}
+
+	expectedExtensiveConfig = &Config{
+		Secret: AclTokenSecret{
+			Provider: "secrets-manager",
+			Configuration: SecretConfiguration{
+				Prefix:                     "abc123",
+				ConsulClientTokenSecretARN: "some-long-arn",
+			},
+		},
+		Mesh: Mesh{
+			BootstrapDir:         "/consul/",
+			HealthSyncContainers: []string{"frontend"},
+			Service: ServiceRegistration{
+				Name:              "frontend",
+				Tags:              []string{"frontend"},
+				Port:              8080,
+				Address:           "127.0.0.1",
+				EnableTagOverride: true,
+				Meta: map[string]string{
+					"env":     "test",
+					"version": "x.y.z",
+				},
+				Weights: &AgentWeights{
+					Passing: 6,
+					Warning: 5,
+				},
+				Checks: []AgentServiceCheck{
+					{
+						CheckID: "frontend-http",
+						Name:    "frontend-http",
+						HTTP:    "http://localhost:8080/health",
+						Method:  "POST",
+						Body:    "{\"method\": \"health\"}",
+						Notes:   "Health check for the frontend service",
+						Header: map[string][]string{
+							"Content-Type": {"application/json"},
+						},
+						Interval:               "30s",
+						Timeout:                "10s",
+						SuccessBeforePassing:   3,
+						FailuresBeforeCritical: 4,
+					},
+					{
+						CheckID:  "frontend-tcp",
+						Name:     "frontend-tcp",
+						TCP:      "localhost:8080",
+						Interval: "15s",
+						Timeout:  "5s",
+					},
+					{
+						CheckID:    "frontend-grpc",
+						Name:       "frontend-grpc",
+						GRPC:       "localhost:8080",
+						GRPCUseTLS: true,
+						Interval:   "20s",
+						Timeout:    "5s",
+					},
+					{
+						CheckID: "frontend-ttl",
+						Name:    "frontend-ttl",
+						TTL:     "10m",
+						Status:  "passing",
+					},
+					// TODO: api.AgentServiceCheck has no H2Ping field (in v1.10.1)
+					// 		 Looks like it's been added to the api package on Consul main,
+					// 		 so should come in a future release.
```

Comment:
I noticed this after correcting the schema to disallow additional properties for checks. `H2Ping` doesn't exist in the Consul `api` package, so users can't register an HTTP2 check (but they couldn't before file-based config either).